3d-n-body-pendulum made a bit nicer. Should fix #534

Peter230655 opened the pull request:
- use ``generate_ode_function`` instead of lambdify
- use ``PyDY Visualization`` to create the 3D animation
- make it more conformant to PEP8

I could not find a simple way to check the conformity using flake8. The original ipynb files were checked so I hope most is o.k.

Peter230655 committed:
3d-n-body-pendulum made a bit nicer

Peter230655 committed:
changes / improvements as per suggestions

Peter230655 committed:
More comments pulled from code to text

Peter230655 committed:
save before merging

Peter230655 committed:
Merge remote-tracking branch 'upstream/solve-ivp-arg-order' into 3d-n-link-pendulum

Peter230655 committed:
use symjit as generator to avoid np.ascontiguoudarray(..)

Peter230655 committed:
added time_first keyword. Integrated rhs_gen into solve_ivp

Peter230655 committed:
just a check if the error persists

Peter230655 committed:
replaced generator='symjit' with 'cython'

Peter230655 committed:
locally it runs without error

Peter230655 committed:
Merge branch 'master' into 3d-n-link-pendulum

Peter230655 commented:
I do not understand the error: it looks like a clear python error.
- it runs fine locally, with make html
- if it is a python error in my program, why does it not fail with every test?
- the number it gives for out of bounds was different in my previous attempt to push it.

Peter230655 committed:
defined schritte again

Peter230655 committed:
re-set times

Peter230655 commented:
No idea, why it passed now.
Maybe some variables get over-written during the test run.

moorepants commented:
Merge in the master branch and let this run again in CI after that.

moorepants commented on docs/examples/3d-n-body-pendulum.rst:
```diff
@@ -8,247 +8,389 @@
 
     t_span = (0., intervall)
 
     def gradient(t, y, args):
```
You can now swap the order of the arguments with #546.

Peter230655 committed:
Merge branch 'master' into 3d-n-link-pendulum

Peter230655 commented:
I will fix the C-contiguous stuff once the force_c_contiguous is merged. So no need to look at it now.

Peter230655 committed:
Merge branch 'master' of https://github.com/Peter230655/pydy into 3d-n-link-pendulum

Peter230655 committed:
added force_c_contiguous=True

Peter230655 commented:
1. added the ``force_c_contiguous=True``. I thought this was the default, but I was wrong
2. Corrected wordings accordingly.
3. Used kane.kindiffdict for the kinematic equations needed in generate_ode_function

I noticed this;
Before solve_ivp integrated without giving values for rtol, atol. Now it needed them, else the integration aborted.

moorepants commented:
Please use System() for the examples. Try to copy the style of the Carvallo-Whipple Bicycle example.

Peter230655 commented:
> Please use System() for the examples. Try to copy the style of the Carvallo-Whipple Bicycle example.

This will take me a while, I have never used it. I guess no rush.

Peter230655 committed:
converted to using System

Peter230655 commented:
Converted the simulation to System, no substantial further changes.

moorepants commented on docs/examples/3d-n-body-pendulum.rst:
```diff
@@ -8,469 +8,568 @@
+    for i in range(resultat.shape[0]):
+        for j, key in enumerate(sys.initial_conditions.keys()):
+            sys.initial_conditions[key] = resultat[i, j]
+        RHS[i] = rhs_gen()
```
This is not the intended use of `System.evaluate_ode()` and results in code we do not want to encourage in an example.

Peter230655 replied:
I will change as soon as I can figure out the signature of evaluate_ode_function. I believe that this would be the correct way.

moorepants replied:
I would leave out calculation of the accelerations in this example. OR we need to add the functionality to System to do this cleanly.

moorepants replied:
I opened this issue: https://github.com/pydy/pydy/issues/570

Peter230655 replied:
> I would leave out calculation of the accelerations in this example. OR we need to add the functionality to System to do this cleanly.

Would it not be useful to have an example where we calculate the reaction forces?
In this example I cannot get the $\dot q$ symbolically with $\dot q =$ kane.rhs(), it takes too long.

Maybe better wait until #570 is available ?

NB: Same point with explorer1_anomaly, except there I had to remove the reaction forces from the force vector.

moorepants replied:
The purpose of PyDy is not to teach different dynamics analysis methods, so no I prefer we keep clean examples that show how to use PyDy's features (same goes for opty).

If there aren't clean ways to do certain common dynamics calculations, then feature requests should be made for that (in sympy, pydy, or opty).

Peter230655 replied:
> The purpose of PyDy is not to teach different dynamics analysis methods, so no I prefer we keep clean examples that show how to use PyDy's features (same goes for opty).
> 
> If there aren't clean ways to do certain common dynamics calculations, then feature requests should be made for that (in sympy, pydy, or opty).

But would #570 not be a clean way?

moorepants replied:
Yes, that's why I proposed adding the feature to meet your need.

Peter230655 replied:
> Yes, that's why I proposed adding the feature to meet your need.

Then I would let the 3d pendulum and the explorer1_anomaly1 sit in waiting, until I can incorporate this feature.
I think, I would be good to show how to calculate reaction forces when to system is too large for kane.rhs() to finish in good time.

moorepants replied:
Another option is that you try to add the features.

Peter230655 commented:
I could try essentially what I did the ugly way with ``evaluate_ode`` inside System.
So, the user would have a method, say,  ``sys.give_qdot(solution), where solution is an n x m array n = len(times), m = len(initial_conditions) and it returns an n array containing the qdot.

Such an approach I can try, but is it not a clumsy one?

moorepants commented:
It is not clumsy, but in #570 I've stated the API we should impelment for this. Not `.give_qdot`.

Peter230655 committed:
Merge branch 'master' into 3d-n-link-pendulum

Peter230655 committed:
use evaluate_ode(x=result, t=sys.times) to get the accelerations

Peter230655 commented:
I used the new feature of ``evaluate_ode(..) to get the ode values needed for the accelerations

moorepants commented on docs/examples/3d-n-body-pendulum.rst:
```diff
@@ -8,469 +8,562 @@
-    iYY1 = iXX1
-    iZZ1 = iXX1
+    sys.times = np.linspace(0., 5.0, 500)
+    times = sys.times   # for later use
```
Just use `sys.times` for later use.

Peter230655 replied:
> Just use `sys.times` for later use.

This was just laziness on my part, I had times in the program. Will correct it.

moorepants commented on docs/examples/3d-n-body-pendulum.rst:
```diff
@@ -8,469 +8,562 @@
 .. jupyter-execute::
 
-    # plot the energies
+    RHS = sys.evaluate_ode(x=resultat, t=sys.times)
```
```suggestion
    RHS = sys.evaluate_ode(x=resultat)
```

no need to pass sys.times, it is already known to system

Peter230655 replied:
> no need to pass sys.times, it is already known to system

Clear, but why is it optional?

moorepants commented on docs/examples/3d-n-body-pendulum.rst:
```diff
@@ -8,469 +8,562 @@
+                        method='Radau',
+                        atol=1.e-8,
+                        rtol=1.e-8,
+                        **kwargs)
```
The intended design is more like this:

```python
def ode_solver(f, x0, ts, args=(), **kwargs):
    return solve_ivp(lambda t, x: f(x, t, *args), ts[[0, -1]], x0, t_eval=ts, **kwargs).y.T

sys = System(..., ode_solver=ode_solver)

sys.integrate(method='Radau', rtol=1e-5)
```

Then you can integrate with whatever method or settings you want with calls to `.integrate()`

Peter230655 replied:
> The intended design is more like this:
> 
> ```python
> def ode_solver(f, x0, ts, args=(), **kwargs):
>     return solve_ivp(lambda t, x: f(x, t, *args), ts[[0, -1]], x0, t_eval=ts, **kwargs).y.T
> 
> sys = System(..., ode_solver=ode_solver)
> 
> sys.integrate(method='Radau', rtol=1e-5)
> ```
> 
> Then you can integrate with whatever method or settings you want with calls to `.integrate()`

Looks nice. I just did it as you had shown me earlier.

Peter230655 committed:
changed as per suggestions

Peter230655 commented:
I will do the same changes with explorer1 anomaly

moorepants commented:
This took 4 minutes to run. I have an open issue https://github.com/pydy/pydy/issues/470 about this.

The examples should, in general, run very fast. Most should run in under 10 seconds with an occasional at a max of like 30 seconds.

Peter230655 commented:
Running it locally takes 52.3 sec, 
I will see what I can do.

moorepants commented:
The 4 minutes is what it takes to run on my computer. Keep in mind that running `make html` has to execute all examples!

Peter230655 commented:
I got it down to 11 sec on my PC. Should I push it?

Peter230655 commented:
> The 4 minutes is what it takes to run on my computer. Keep in mind that running `make html` has to execute all examples!

I noticed this, but how do I see how long this takes locally, other than using a stopwatch?

Peter230655 committed:
reduced running time to 11.5 sec from around 53 sec

moorepants commented:
I've thought a bit more about this example and my general opinion is that is just too complex (both in the amount/type of things you want to show and the complexity of the code itself).

We already have a conical pendulum example. It is one of 4 simple (redundant) holonomic systems we already have in the sphinx documentation. The examples should be short and sweet, in general, and show off features of PyDy (not limitations).

This example adds two things not seen in the other sphinx examples: energy calculations and noncontributing forces. System() cannot handle either of these and there are two long open issues about solving these: #136 and #127. We should solve those issues before adding examples that need such features.

Even if we have this example as designed, there are simpler ways to do both energy and noncontributing forces that only require minor abuses of the intended use design of PyDy. For energy, you can write the power equation and add it as an extra differential equation, then energy is simply returned by the ode solver. For noncontributing forces, you can do something similar, treating the noncontributing force measure numbers as a state derivative. Then the right hand side function will return the force components. Both of these require using the ODE generators directly because System() populates the generators automatically, ignoring the auxiliary equations or any other equations present. But it is ideal that the examples use System(), not the ODE generators. We can explain the generators with very simple examples that don't even require making a model from scratch.

I think that the reaction forces should be removed from the example. Plotting energy is not too complex, so it can stay. But the code needs to be neat and simple, like the other examples. It is difficult to learn from complex code. I've mentioned on the opty repo that looking at the examples shown in matplotlib and scikit-learn galleries give a good idea of how simple documentation examples should be. We tend to let examples get out of hand (both in complexity for learning and in computational time needed to execute them and in redundancy). We also have an open issue about dealing with the examples we already have #422, yet you are adding new examples, instead of trying to showcase the ones we already have made.

Your general approach, which you are very talented at, is to make elaborate work-arounds given the limitations of the tools rather than improve the tool. If you want PyDy to do something it doesn't already do, then the ideal approach is to help improve PyDy. You usually have to make a workaround one or more times in your own code to help see the more general issue, but we should limit how much we show elaborate work-arounds in examples.

This particular example is already present in PyDy and the open issues about it are #534 and #470, so the goal should be to simplify both so it works with PyDy and runs fast.

Peter230655 commented:
> I've thought a bit more about this example and my general opinion is that is just too complex (both in the amount/type of things you want to show and the complexity of the code itself).
> 
> We already have a conical pendulum example. It is one of 4 simple (redundant) holonomic systems we already have in the sphinx documentation. The examples should be short and sweet, in general, and show off features of PyDy (not limitations).

I totally agree, that it is not needed for the reasons you mentioned. The only reason I played around with it is that it already was an example in pydy, it was very ugly and I wanted to make it nicer.
> 
> This example adds two things not seen in the other sphinx examples: energy calculations and noncontributing forces. System() cannot handle either of these and there are two long open issues about solving these: #136 and #127. We should solve those issues before adding examples that need such features.

explorer1 anomaly also has these features, and at least it is not a pendulum. I think, better to show these features.
I will try the energy as you explained below, and if you can explain thereaction forces idea, I will try it there, too.
> 
> Even if we have this example as designed, there are simpler ways to do both energy 

The kinetic energy is very simple right now: kinetic_energy = sum([body.kinetic_enery(N) for body in body_list])
The spring enery and the potential energy are a bit more involved.


and noncontributing forces that only require minor abuses of the intended use design of PyDy. For energy, you can write the power equation and add it as an extra differential equation, then energy is simply returned by the ode solver. For noncontributing forces, you can do something similar, treating the noncontributing force measure numbers as a state derivative. Then the right hand side function will return the force components. Both of these require using the ODE generators directly because System() populates the generators automatically, ignoring the auxiliary equations or any other equations present. But it is ideal that the examples use System(), not the ODE generators. We can explain the generators with very simple examples that don't even require making a model from scratch.

I understand the idea with the energy: energy = $\int_0^{t_f} power(s) ds$, and I will try this.
I am not clear about the noncontributing forces. Say, I use $f_x, f_y, f_z$ as the noncontributing forces componemts when I set up Kane's equations. Then would I do something like $\frac{d}{dt} react_{f_x} = f_x$ ?
> 
> I think that the reaction forces should be removed from the example. Plotting energy is not too complex, so it can stay. But the code needs to be neat and simple, like the other examples. It is difficult to learn from complex code. I've mentioned on the opty repo that looking at the examples shown in matplotlib and scikit-learn galleries give a good idea of how simple documentation examples should be. We tend to let examples get out of hand (both in complexity for learning and in computational time needed to execute them and in redundancy). We also have an open issue about dealing with the examples we already have #422, yet you are adding new examples, instead of trying to showcase the ones we already have made.
> 
> Your general approach, which you are very talented at, is to make elaborate work-arounds given the limitations of the tools rather than improve the tool. If you want PyDy to do something it doesn't already do, then the ideal approach is to help improve PyDy. You usually have to make a workaround one or more times in your own code to help see the more general issue, but we should limit how much we show elaborate work-arounds in examples.

My programmiing skills are just too limited to improve a library. I can test new features once available.
> 
> This particular example is already present in PyDy and the open issues about it are #534 and #470, so the goal should be to simplify both so it works with PyDy and runs fast.

As we already have pendulums in the examples, I think best to simply **remove this one**. The version we have is complex _and_ ugly, and now it is less ugly and likely more complex.

moorepants commented:
> My programmiing skills are just too limited to improve a library.

You may think that, but I think you certainly have enough skills.

>  I think best to simply remove this one. 

I'm fine keeping it, but just simplify it. For example, remove the reaction forces entirely and we merge. Then you can open an issue here (or tag onto the one about lagrange) with the request to support aux equations. Show this as an example of how there is a limitation.

Peter230655 commented:
> > My programmiing skills are just too limited to improve a library.
> 
> You may think that, but I think you certainly have enough skills.

Thanks for the compliment, but my skills really are limited, at least as soon as classe, base classes, etc get involved.
> 
> > I think best to simply remove this one.
> 
> I'm fine keeping it, but just simplify it. For example, remove the reaction forces entirely and we merge. Then you can open an issue here (or tag onto the one about lagrange) with the request to support aux equations. Show this as an example of how there is a limitation.

I will remove the reaction forces and open an issue about support for the auxiliary equations.
I will try the energies using your idea above.

moorepants commented:
> I am not clear about the noncontributing forces. 

If the state time derivative vector is: [q', u', j'] where j is the impulse (the integral of the force wrt time), then you can just add impulse as a new state variable. That means solve_ivp will return the impulse versus time and then rhs() will return the force versus time.

Peter230655 committed:
removed reaction forces and angular momentum

Peter230655 commented:
I removed the noncontributing forces and the angular momentum which is of little interest here.
As I am not yet fully clear how to get the energy from the power, I left this as it is. I need to play around with such ideas for a whule, until I feel confident.

Peter230655 commented:
> I've thought a bit more about this example and my general opinion is that is just too complex (both in the amount/type of things you want to show and the complexity of the code itself).
> 
> We already have a conical pendulum example. It is one of 4 simple (redundant) holonomic systems we already have in the sphinx documentation. The examples should be short and sweet, in general, and show off features of PyDy (not limitations).
> 
> This example adds two things not seen in the other sphinx examples: energy calculations and noncontributing forces. System() cannot handle either of these and there are two long open issues about solving these: #136 and #127. We should solve those issues before adding examples that need such features.
> 
> Even if we have this example as designed, there are simpler ways to do both energy and noncontributing forces that only require minor abuses of the intended use design of PyDy. For energy, you can write the power equation and add it as an extra differential equation, then energy is simply returned by the ode solver. For noncontributing forces, you can do something similar, treating the noncontributing force measure numbers as a state derivative. Then the right hand side function will return the force components. Both of these require using the ODE generators directly because System() populates the generators automatically, ignoring the auxiliary equations or any other equations present. But it is ideal that the examples use System(), not the ODE generators. We can explain the generators with very simple examples that don't even require making a model from scratch.
> 
> I think that the reaction forces should be removed from the example. Plotting energy is not too complex, so it can stay. But the code needs to be neat and simple, like the other examples. It is difficult to learn from complex code. I've mentioned on the opty repo that looking at the examples shown in matplotlib and scikit-learn galleries give a good idea of how simple documentation examples should be. We tend to let examples get out of hand (both in complexity for learning and in computational time needed to execute them and in redundancy). We also have an open issue about dealing with the examples we already have #422, yet you are adding new examples, instead of trying to showcase the ones we already have made.

I looked at docs/examples/rocket_car. I could not really see, how it shows the capabilities of PyDy. I looked to me like the solution of a very difficult exam you gave at the end of your lecture.. :-) 
I could convert this to an example to show PyDy's capabilities - but it would not be the same anymore: I'd use sympy.physics.mechanics to get ``kane`` for System(), and then numerically integrate it.
Is this what should be done?

> 
> Your general approach, which you are very talented at, is to make elaborate work-arounds given the limitations of the tools rather than improve the tool. If you want PyDy to do something it doesn't already do, then the ideal approach is to help improve PyDy. You usually have to make a workaround one or more times in your own code to help see the more general issue, but we should limit how much we show elaborate work-arounds in examples.
> 
> This particular example is already present in PyDy and the open issues about it are #534 and #470, so the goal should be to simplify both so it works with PyDy and runs fast.

moorepants commented:
> I looked at docs/examples/rocket_car. I could not really see, how it shows the capabilities of PyDy. I looked to me like the solution of a very difficult exam you gave at the end of your lecture.. :-)

That was a student that was eager to contribute and he wanted to figure out variable mass. The two examples he added were well formatted and had free body diagrams, so I merged them. To get new contributors, the bar can't be too high. He didn't hang around past those two examples though. Many examples here are also historical and even existed before this PyDy repo and the current incarnation with System() and the ode generators.

I don't want to make changes to existing examples in the sphinx documentation, most are fine and clear. We have legacy examples (some pre-PyDy) that could be improved as sphinx examples.

It would be nice if we could just dump any and all examples here, but it quickly gets unmanagemable and there isn't a clear purpose is showing the software.

This is how I currently see things:

The software stack is made up of SymPy.vector/mechanics/biomechanics, Learn Multibody Dynamics, PyDy, and opty.

- SymPy: should have examples that are primarily symbolic and demonstrate the features of the parts of SymPy.
- LMD: book focused on teaching principles of dynamics using SymPy mechanics.
- PyDy: should show how to use the symbolic-bridge, simulation, and visualization
- opty: should show how to solve optimal control problems

The documentation examples (in all repositories) should run fast. They should, in general, be simple and highlight features of the software. Ideally the examples are the minimal complexity needed to show a feature.

In the past, this repo and somewhat sympy was a clearing house for any example people made. But I do not think that is the ideal approach because we end up with many broken examples that don't run or are not in a format that is easy to learn from. I don't really want any more of that, we have enough already.

The problems you tend to solve are nontrivial and belong somewhere in between simple documentation examples and a research paper. That's why we created pst-notebooks, so you can drop all your examples there as you see fit.

I have not had time to make some kind of clear guidelines on what is and isn't acceptable as an example. It isn't even that clear in my head, because I haven't written it down. So you are left guessing a bit about this. I have tried to explain it to you numerous times in the opty repo and elsewhere, even in person.

I'll once again point to matplotlib's gallery:

https://matplotlib.org/stable/gallery/index.html

I think this is a good model for documentation. It is focused on demonstrating features of matplotlib in minimal, simple examples. Most real matplotlib code is big and messy, but they don't have that here in this gallery. That is what we should strive for in the documentation.

SymPy's documentation is not so well organized. SymPy lags behind these other projects on many things like this, as we are less organized. So I wouldn't necessarily copy SymPy's approach, but even there we have started to move examples into "tutorials" and learning prose into "explanations". There is more work to be done there to further improve that.

Bottom line, I'm not so interested in merge pst-notebook style examples into sympy, pydy, or opty. They are just too complex, take too long to run, and it is hard to see the forest for the trees for learning/documentation purposes.

In PyDy, I think the model examples we have in the sphinx docs are: mass/spring/damper, bicycle model, holonomic spring pendulum. They all have a free body diagram (svg) and plenty of text explaning each step and avoid lots of complexities. I'm open to new examples that follow that structure/format/style which highlight things about PyDy we have not highlighted well in other places. We can also improve those examples to show extra features without having to add entirely new examples (for some features).

The goal in the documentation of all these packages is not focused on demonstrating the maximal capabilities of the software, or the limits of types of problems it can handle, or to teach anything about dynamics. For the future, we should tend towards example gallery's like matplotlib's.

moorepants commented:
The rocket car example would better belong in SymPy, since it is only symbolic.

Peter230655 commented:
> > I looked at docs/examples/rocket_car. I could not really see, how it shows the capabilities of PyDy. I looked to me like the solution of a very difficult exam you gave at the end of your lecture.. :-)
> 
> That was a student that was eager to contribute and he wanted to figure out variable mass. The two examples he added were well formatted and had free body diagrams, so I merged them. To get new contributors, the bar can't be too high. He didn't hang around past those two examples though. Many examples here are also historical and even existed before this PyDy repo and the current incarnation with System() and the ode generators.

Now very clear, without this explanation it was not obvious why the example was included.
> 
> I don't want to make changes to existing examples in the sphinx documentation, most are fine and clear. We have legacy examples (some pre-PyDy) that could be improved as sphinx examples.
> 
> It would be nice if we could just dump any and all examples here, but it quickly gets unmanagemable and there isn't a clear purpose is showing the software.
> 
> This is how I currently see things:
> 
> The software stack is made up of SymPy.vector/mechanics/biomechanics, Learn Multibody Dynamics, PyDy, and opty.
> 
> * SymPy: should have examples that are primarily symbolic and demonstrate the features of the parts of SymPy.
> * LMD: book focused on teaching principles of dynamics using SymPy mechanics.
> * PyDy: should show how to use the symbolic-bridge, simulation, and visualization
> * opty: should show how to solve optimal control problems
> 
> The documentation examples (in all repositories) should run fast. They should, in general, be simple and highlight features of the software. Ideally the examples are the minimal complexity needed to show a feature.

This is an advantage how you set up pst-notebooks: running time does not seem to matter.
> 
> In the past, this repo and somewhat sympy was a clearing house for any example people made. But I do not think that is the ideal approach because we end up with many broken examples that don't run or are not in a format that is easy to learn from. I don't really want any more of that, we have enough already.
> 
> The problems you tend to solve are nontrivial and belong somewhere in between simple documentation examples and a research paper. That's why we created pst-notebooks, so you can drop all your examples there as you see fit.

Right! It also has the advantage that running time does not matter. I believe the latest additions to PyDy (calculate the constraint values) are not released, yet. As soon as you release them I will push my rattleback to pst-notebooks.
> 
> I have not had time to make some kind of clear guidelines on what is and isn't acceptable as an example. It isn't even that clear in my head, because I haven't written it down. So you are left guessing a bit about this. I have tried to explain it to you numerous times in the opty repo and elsewhere, even in person.

I think, I understand the guidelines, but then an example like the rocket car throws me off. Now clear.
> 
> I'll once again point to matplotlib's gallery:

Really perfect! I often use chatGPT to make 3D animations, where I don't care much how they work as long as they do. It likely gets its wisdom from there.
> 
> https://matplotlib.org/stable/gallery/index.html
> 
> I think this is a good model for documentation. It is focused on demonstrating features of matplotlib in minimal, simple examples. Most real matplotlib code is big and messy, but they don't have that here in this gallery. That is what we should strive for in the documentation.
> 
> SymPy's documentation is not so well organized. SymPy lags behind these other projects on many things like this, as we are less organized. So I wouldn't necessarily copy SymPy's approach, but even there we have started to move examples into "tutorials" and learning prose into "explanations". There is more work to be done there to further improve that.
> 
> Bottom line, I'm not so interested in merge pst-notebook style examples into sympy, pydy, or opty. They are just too complex, take too long to run, and it is hard to see the forest for the trees for learning/documentation purposes.

I have absolutely no problem with this! For me the fun is getting them to work. Whether they are published anywhere is secondary to me.
> 
> In PyDy, I think the model examples we have in the sphinx docs are: mass/spring/damper, bicycle model, holonomic spring pendulum. They all have a free body diagram (svg) and plenty of text explaning each step and avoid lots of complexities. I'm open to new examples that follow that structure/format/style which highlight things about PyDy we have not highlighted well in other places. 

>We can also improve those examples to show extra features without having to add entirely new examples (for some features).

This is what I tend to forget :-(
E.g. surely there is an existing example, where the new "calculate constraint values" could be added.
> 
> The goal in the documentation of all these packages is not focused on demonstrating the maximal capabilities of the software, or the limits of types of problems it can handle, or to teach anything about dynamics. For the future, we should tend towards example gallery's like matplotlib's.

I think, with my existing PRs we should proceed as follows:
1. This example (3d-n-body-pendulum was in pydy examples before I 'beautified' it. I do not mind at all, if you delete the existing old version. We do not need another pendulum. If you do not delete it, this one is better than the old one.
2. I will close explorer1-anomaly and move it to pst-notebooks. 
3. I will close the rattleback and move it to pst-notebooks, as soon as you release PdDy with the latest features.

Peter230655 commented:
> The rocket car example would better belong in SymPy, since it is only symbolic.

Would it be too complicated for LMD?

moorepants commented:
> Would it be too complicated for LMD?

I do not currently have any plans to add anything about variable mass systems to my book.

moorepants commented:
If you want to add the noncontributing forces back, you can. Use the new feature of System.

Peter230655 commented:
I am out for the weekend. Will try as soon as I am back.

moorepants commented:
There are no deadlines in volunteer open source. Enjoy your weekend.

Peter230655 committed:
Merge branch 'master' into 3d-n-link-pendulum

Peter230655 commented:
I tried to implement the new feature to calculate the noncontributing forces.
- I added the virtual speeds and reaction forces in Kane's equations
- I gave the list: sys.concontributing_forces = [fx, fy, fz]
- I thought the last three colums of the result are these forces so this is what the plot shows.

As comparison I also calculated them the old way, using KM,auxiliary_eqs.
The results are very different. 
In particular the new way gives fx(t=0) = fy(t=0) = fz(t=0) = 0.000 
Intuitively this cannot be correct, after all the pendulum is hanging on the suspension point P[0].
The old way gives 'reasonable' values.

When I tried to use sys.generate_ode_function(generator='cython'...) I got an error ``module could not be loaded`` or similar.
This feature worked before in my environment. No idea why it fails now.

moorepants commented:
I don't see anything about noncontributing forces here, so I can't really help.

Peter230655 commented:
> I don't see anything about noncontributing forces here, so I can't really help.

I see that the file changed does not show anything. No idea what mistake I made. I will try again.

Peter230655 committed:
new attempt

Peter230655 commented:
Now files changed seems to look correct.

moorepants commented:
It looks like you are plotting the time integral of the forces, not the forces.

Peter230655 commented:
> It looks like you are plotting the time integral of the forces, not the forces.

This was the problem, now it looks like the old method.
Do you have any idea bout my ``generator=cython`` problem?

moorepants commented:
I can't help you debug with zero information. You have to provide either code I can run or output of your error messages or something. You can also enable debugging with cython using the flags to the generator. I would not use the cython generator for this problem.

moorepants commented:
Some general comments on this:

- It would make the example clearer if all of the loops related to extending it beyond 3 links were removed.
- Please move all energy expressions to the new outputs feature (or anything else that you calculate).

moorepants commented:
> This was the problem, now it looks like the old method.

All this is explained in the documentation. I tried to make it clear. You told me it was fine in #583.

Peter230655 commented:
> Some general comments on this:
> 
>     * It would make the example clearer if all of the loops related to extending it beyond 3 links were removed.

Will do so. I agree this clouds the real things a bit.
> 
>     * Please move all energy expressions to the new outputs feature (or anything else that you calculate).

Makes emminent sense. I did no think of this!

Peter230655 commented:
> > This was the problem, now it looks like the old method.
> 
> All this is explained in the documentation. I tried to make it clear. You told me it was fine in #583.

It IS fine! I just did not think. If there is a method ``evaluate_outputs`` it makes sense to use it to evaluate outputs!

Peter230655 commented:
> I can't help you debug with zero information. You have to provide either code I can run or output of your error messages or something. You can also enable debugging with cython using the flags to the generator. I would not use the cython generator for this problem.

Thanks! I will revert to this issue once I am done with the pendulum.

Peter230655 commented:
The energy using sys.outputs seems to work fine.
Right now I just set the speed of the suspension point equal to zero.
I could use configuration / speed constraints instead, then this simulation would show all the new capabilities.
Danger of overloading?

moorepants commented:
>  Danger of overloading?

Yes.

Peter230655 committed:
changed as per suggestions

Peter230655 commented:
Calculate the energies using sys,outputs.
Calculate the noncontributing forces using the new capability.

moorepants commented:
Ok, major improvement now. Merging.
